spec/ics-2-consensus-verification/README.md

```diff
@@ -0,0 +1,219 @@
+      in `[Message]` 
+      // XXX: make it readable
+
+* `update` is a helper function using `Consensus.verify` to update the existing 
+  `RootOfTrust`, defined as
+
+```
+function update(cons :: Consensus, rot :: RootOfTrust, h :: Header) returns (Error|RootOfTrust) {
+    if cons.verify(rot, h):
+        if h.base != Nothing:
+            return RootOfTrust(h.base, h.root)
+        else:
+            return RoofOfTrust(rot.base, h.root)
+    else:
+        return Error  
+   
+}
+``` 
+
+##### Functions
+
+* `Height :: Blockchain -> Header -> Uint` returns the position of the header in the
+  Blockchain's header list.
+
+#### Requirements
+
+Consensus verification requires the following accumulator primitives with datastructures and
+properties as defined in ICS23:
+
+* `AccumulatorRoot`
+
+#### Subprotocols
+
+##### Verifier
+
+The verifier algorithm checks that new `Header`s were in fact generated by a blockchain. It is expected to verify a `Header` 
+efficiently; far more efficiently than replaying `Consensus` logic for the given parent `Header` and the
+list of network messages, idealy in O(1) time. `Header.p` provides the proof that the verifier can use. 
+Verifiers assume the following properties will be satisfied for the `Header`s:
+
+1. `Header`s have no more than one direct child
+ 
+* Satisfied if: deterministic safety
+* Possible violation scenario: validator double signing, chain reorganization (Nakamoto consensus)
+
+2. `Header`s eventually have at least one direct child
+
+* Satisfied if: liveness, light-client verifier continuity
+* Possible violation scenario: synchronized halt, incompatible hard fork
+
+3. `Header`s are generated from the `Consensus`, which ensures valid transition of the state
+
+* Satisfied if: correct block generation & state machine
+* Possible violation scenario: invariant break, validator cartel
+
+##### Accumulator Root
+
+`RootOfTrust` contains an accumulator root, which identifies the whole state of the 
+corresponding blockchain at the point of time that the commit is generated. It is expected that 
+the verifying inclusion or exclusion of certain data in the accumulator can be done efficiently. See 
+ICS23 for the details about the `AccumulatorRoot`s.
+
+##### Consensus 
+
+`Consensus` is a blockchain protocol which actually generates a list of `Header`s from the latest
+state and the incoming transactions. While the chains on the network does not directly proving the 
+consensus process, it is expected that the consensus algorithms will generate valid `Header`s.
+
+### Example Implementation
+
+An example blockchain `B` runs on a single operator consensus algorithm, called `Op`. If a 
+block is signed by the operator, then it is valid. The operator signing key can be changed while
+the chain is running. In that case, the new header stores the updated pubkey. 
+
+`H` contains `LogStore`, which is a list with type of `[bytes]`. The whole state is serialized 
```

Comment:
Not clear what LogStore is...

Reply:
I will change the example to a simple KVStore, which seems more appropriate